### PR TITLE
React hook signature & add promises to hook action results

### PR DESCRIPTION
**Pull request by plmercereau**
## Post-merge to-do list
- [ ] publish
- [ ] once published, bump versions in example, and remove the deprecated syntax in the react-apollo example

## Recap
1. deprecate the following syntax
```js
const { signInEmailPasswordless } = useSignInEmailPasswordless('nuno@fcporto.pt')
signInEmailPasswordless()
```
Instead, use
```js
const { signInEmailPasswordless } = useSignInEmailPasswordless()
signInEmailPasswordless('nuno@fcporto.pt')
```
-> no breaking change has been introduced. For instance, `useSignUpEmailPassword('szilard@brussels.be','1234', options)` will appear as deprecated but will work, while `useSignUpEmailPassword(options)` will work too.

2. Return a promise in hook actions:
```js
const { signInEmailPasswordless } = useSignInEmailPasswordless()
const { isSuccess, isError, error } = await signInEmailPasswordless('johan@ikea.se')
```
3. Sync up the states (the re-rendering problem we talked about between isSuccess, isError, error etc: cannot reproduce
4. Documentation and examples of the above

### CODE REVIEW

File: examples/react-apollo/src/components/email-passwordless-form.tsx [hunk 1]
```diff
@@ -6,9 +6,13 @@ import React, { useState, useEffect } from 'react'
...
-    redirectTo: '/profile'
-  })
+  const { signInEmailPasswordless, isError, isSuccess, error } = useSignInEmailPasswordless(
+    // TODO correct this once the new packages are released
+    undefined,
+    {
+      redirectTo: '/profile'
+    }
+  )
   const [showError, setShowError] = useState(true)
```
Comment (elitan) [inline examples/react-apollo/src/components/email-passwordless-form.tsx:16]: is this `showError` needed?

File: packages/react/src/hooks/user.ts [hunk 2]
```diff
@@ -7,64 +7,171 @@ import {
...
+        type: 'REQUEST',
+        email: typeof valueEmail === 'string' ? valueEmail : stateEmail,
+        options: valueOptions
+      })
+      service.onTransition((state) => {
+        if (state.matches({ idle: 'error' })) {
+          const error = state.context.error
+          setIsError(true)
+          setError(error)
+          setIsLoading(false)
```
Comment (elitan) [inline packages/react/src/hooks/user.ts:64]: Is this fine now? It won't cause 3 re-renders?

File: examples/react-apollo/src/components/email-passwordless-form.tsx (hunk 1 above)
Comment (plmercereau) [reply examples/react-apollo/src/components/email-passwordless-form.tsx:16]: yes, this is required to hide the error message when the user changes the email field

Comment (elitan) [reply examples/react-apollo/src/components/email-passwordless-form.tsx:16]: Ok, I would maybe not bother with that but I see your point.

File: examples/react-apollo/src/sign-up/email-password.tsx [hunk 3]
```diff
@@ -15,8 +15,9 @@ export const EmailPassword: React.FC = () => {
   const navigate = useNavigate()
   const [confirmPassword, setConfirmPassword] = useState('')
   const { signUpEmailPassword, error, needsEmailVerification, isSuccess } = useSignUpEmailPassword(
-    email,
-    password,
+    // TODO change once the new packages have been release
+    undefined,
+    undefined,
     options
```
Comment (elitan) [inline examples/react-apollo/src/sign-up/email-password.tsx:21]: can we remove this now?

**Review by elitan — APPROVED** [review]

File: examples/react-apollo/src/sign-up/email-password.tsx (hunk 3 above)
Comment (elitan) [reply examples/react-apollo/src/sign-up/email-password.tsx:21]: ignore

**Pull request merged**